### ISSUE TITLE
Evaluate/test core optimizations introduced in CRoaring

### ISSUE THREAD

**Issue by lemire** [RoaringBitmap/RoaringBitmap#107]
There is a companion project called CRoaring (https://github.com/RoaringBitmap/CRoaring) which aims to produce faster lower-level Roaring bitmaps.

As part of this work, two optimizations have been found (tenatively) to have some benefits.

First optimization
----------------------

Currently, as part of the in-place lazy union, when we encounter two array containers, we proceed as in the non-lazy union. That is, we generate a bitset container if the cardinality is above 4096, and otherwise, we generate an array container.

https://github.com/RoaringBitmap/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/Container.java#L554

In the CRoaring code, we now have a lazy arrayxarray union which always generate a bitset if the total cardinality of the inputs exceeds a threshold (1024).

https://github.com/RoaringBitmap/CRoaring/blob/master/src/containers/mixed_union.c#L181-L199

To compensate for this optimization, when calling "repairAfterLazy" on a bitset container, we have to convert the bitset container to an array container if the cardinality is insufficient. The code is very simple...

https://github.com/RoaringBitmap/CRoaring/blob/master/include/containers/containers.h#L73-L79

In some instances (as in Census1881), this optimization *divided* the running time by almost 4... because we replaced expensive array unions with relatively cheap array-bitset unions.

First optimization (part 2)
----------------------------------

Even more aggressively, we can proactively convert the union of two array containers to a bitmap in lazy mode as soon as they exceed a small threshold (1024):
https://github.com/RoaringBitmap/CRoaring/blob/7542918a7c07b2cfdb241b24619a2a078f1760db/src/containers/mixed_union.c#L185

Second optimization
---------------------------

When checking whether a run container is more economical under "run optimize" we always convert even if the run container only saves us a single byte. Instead, we can choose to only convert to a run container if it cuts storage by half...

https://github.com/RoaringBitmap/CRoaring/blob/master/include/containers/perfparameters.h#L16

In theory, this could double the storage requirements, but in practice it brings about a much more modest storage increase, while improving significantly the performance in some cases. So it is a different and maybe more interesting trade-off.

Overall
--------------

Anyhow, these are just experimental ideas. They are both very easy to implement, and not difficult to test.

c.c. @gssiyankai @owenkaser

**Comment by lemire** [comment]
I have implemented the first part of the first optimization.

Let us first consider the case without run optimization.

As it did with CRoaring, it is proving to have a hugely beneficial effect in some cases (as in "it divides the running time".

before:
```
$ cat  pristine_5may2016/wideornaivelog.txt |grep "true           roaring"
RealDataBenchmarkWideOrNaive.wideOr_naive           census-income        false           roaring  avgt    5    2298.686 ±   60.337  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive              census1881        false           roaring  avgt    5    4339.512 ±   17.455  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_008        false           roaring  avgt    5    2303.149 ±    5.331  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_003        false           roaring  avgt    5    8179.985 ±   52.937  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_033        false           roaring  avgt    5    1940.826 ±   32.804  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive            uscensus2000        false           roaring  avgt    5     242.272 ±    0.976  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive         weather_sept_85        false           roaring  avgt    5    2949.284 ±   23.083  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive      wikileaks-noquotes        false           roaring  avgt    5    3913.263 ±   16.920  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive       census-income_srt        false           roaring  avgt    5    1525.604 ±    4.493  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive          census1881_srt        false           roaring  avgt    5    6286.240 ±   24.274  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive     weather_sept_85_srt        false           roaring  avgt    5    1644.513 ±    3.707  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive  wikileaks-noquotes_srt        false           roaring  avgt    5    4554.803 ±   23.244  us/op
```


after:
```
$ cat  wideornaivelog.txt |grep "true           roaring"
RealDataBenchmarkWideOrNaive.wideOr_naive           census-income        false           roaring  avgt    5     724.432 ±     3.301  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive              census1881        false           roaring  avgt    5    4405.000 ±    15.698  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_008        false           roaring  avgt    5    2369.230 ±    18.321  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_003        false           roaring  avgt    5    8166.662 ±    29.913  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_033        false           roaring  avgt    5    1923.320 ±    39.977  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive            uscensus2000        false           roaring  avgt    5     241.903 ±     1.120  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive         weather_sept_85        false           roaring  avgt    5    2947.240 ±    17.818  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive      wikileaks-noquotes        false           roaring  avgt    5    3825.752 ±     9.138  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive       census-income_srt        false           roaring  avgt    5     670.154 ±     3.805  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive          census1881_srt        false           roaring  avgt    5    6242.799 ±    29.656  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive     weather_sept_85_srt        false           roaring  avgt    5    1644.999 ±     5.419  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive  wikileaks-noquotes_srt        false           roaring  avgt    5    4182.568 ±    20.888  us/op
```


However, once we turn out run optimization, it has a much smaller effect (though, if it does have an effect, it is probably beneficial).

Before:
```
RealDataBenchmarkWideOrNaive.wideOr_naive           census-income        false  roaring_with_run  avgt    5     687.974 ±     3.221  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive              census1881        false  roaring_with_run  avgt    5    8894.689 ±    30.280  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_008        false  roaring_with_run  avgt    5    2581.049 ±    10.977  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_003        false  roaring_with_run  avgt    5    1772.724 ±    32.377  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_033        false  roaring_with_run  avgt    5     763.996 ±     5.464  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive            uscensus2000        false  roaring_with_run  avgt    5     243.297 ±     0.659  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive         weather_sept_85        false  roaring_with_run  avgt    5    2925.167 ±    25.377  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive      wikileaks-noquotes        false  roaring_with_run  avgt    5    9627.497 ±    15.178  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive       census-income_srt        false  roaring_with_run  avgt    5      57.825 ±     0.429  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive          census1881_srt        false  roaring_with_run  avgt    5    4898.211 ±     6.997  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive     weather_sept_85_srt        false  roaring_with_run  avgt    5     796.107 ±     0.499  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive  wikileaks-noquotes_srt        false  roaring_with_run  avgt    5    1955.759 ±     5.891  us/op
```

After:
```
RealDataBenchmarkWideOrNaive.wideOr_naive           census-income        false  roaring_with_run  avgt    5     689.739 ±    4.491  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive              census1881        false  roaring_with_run  avgt    5    8864.362 ±   30.672  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_008        false  roaring_with_run  avgt    5    2555.903 ±    6.595  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_003        false  roaring_with_run  avgt    5    1802.401 ±   15.553  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_033        false  roaring_with_run  avgt    5     745.722 ±    5.255  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive            uscensus2000        false  roaring_with_run  avgt    5     249.277 ±    0.682  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive         weather_sept_85        false  roaring_with_run  avgt    5    2927.006 ±   15.805  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive      wikileaks-noquotes        false  roaring_with_run  avgt    5    9894.127 ±   41.808  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive       census-income_srt        false  roaring_with_run  avgt    5      57.646 ±    0.792  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive          census1881_srt        false  roaring_with_run  avgt    5    4980.639 ±   25.732  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive     weather_sept_85_srt        false  roaring_with_run  avgt    5     807.045 ±    0.898  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive  wikileaks-noquotes_srt        false  roaring_with_run  avgt    5    1953.400 ±    6.425  us/op
```

**Comment by lemire** [comment]
Here is the implementation of part 1:

https://github.com/RoaringBitmap/RoaringBitmap/commit/d42b5992bc901cda4748ad6823280662af5352d0

**Comment by lemire** [comment]
The second part of the first optimization seems to help quite a bit too in some cases...

```
$  cat  wideornaivelog.txt |grep "false           roaring"
RealDataBenchmarkWideOrNaive.wideOr_naive           census-income        false           roaring  avgt    5     640.938 ±    4.585  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive              census1881        false           roaring  avgt    5    1309.224 ±    1.473  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_008        false           roaring  avgt    5    2318.794 ±   47.204  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_003        false           roaring  avgt    5    4892.582 ±   87.293  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_033        false           roaring  avgt    5    1879.071 ±   29.928  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive            uscensus2000        false           roaring  avgt    5     243.088 ±    0.687  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive         weather_sept_85        false           roaring  avgt    5    2890.097 ±   49.303  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive      wikileaks-noquotes        false           roaring  avgt    5     653.058 ±    1.279  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive       census-income_srt        false           roaring  avgt    5     654.654 ±    2.912  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive          census1881_srt        false           roaring  avgt    5    1557.849 ±    7.022  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive     weather_sept_85_srt        false           roaring  avgt    5    1592.689 ±   80.414  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive  wikileaks-noquotes_srt        false           roaring  avgt    5     652.217 ±    5.539  us/op
```

and 

```
$  cat  wideornaivelog.txt |grep "false  roaring_with_run"
RealDataBenchmarkWideOrNaive.wideOr_naive           census-income        false  roaring_with_run  avgt    5     686.188 ±    3.091  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive              census1881        false  roaring_with_run  avgt    5    2233.381 ±    6.926  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_008        false  roaring_with_run  avgt    5    2549.240 ±    1.403  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_003        false  roaring_with_run  avgt    5    1787.312 ±   19.681  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_033        false  roaring_with_run  avgt    5     760.223 ±    5.209  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive            uscensus2000        false  roaring_with_run  avgt    5     245.437 ±    0.637  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive         weather_sept_85        false  roaring_with_run  avgt    5    2866.310 ±   52.670  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive      wikileaks-noquotes        false  roaring_with_run  avgt    5    9894.356 ±   15.015  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive       census-income_srt        false  roaring_with_run  avgt    5      57.803 ±    0.952  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive          census1881_srt        false  roaring_with_run  avgt    5    4995.352 ±   19.627  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive     weather_sept_85_srt        false  roaring_with_run  avgt    5     793.142 ±    0.925  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive  wikileaks-noquotes_srt        false  roaring_with_run  avgt    5    1920.696 ±    5.835  us/op
```

The code is nearly trivial : 
https://github.com/RoaringBitmap/RoaringBitmap/commit/1075da9090cec78b10127cff5a509c64187f1626

**Comment by lemire** [comment]
The second optimization seems to bring no benefit...

```
$ cat  wideornaivelog.txt |grep "false  roaring_with_run"
RealDataBenchmarkWideOrNaive.wideOr_naive           census-income        false  roaring_with_run  avgt    5     687.793 ±     3.371  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive              census1881        false  roaring_with_run  avgt    5    2229.742 ±     6.619  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_008        false  roaring_with_run  avgt    5    2555.814 ±     9.357  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_003        false  roaring_with_run  avgt    5    1736.942 ±    44.636  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive           dimension_033        false  roaring_with_run  avgt    5     765.297 ±     4.222  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive            uscensus2000        false  roaring_with_run  avgt    5     250.107 ±     0.655  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive         weather_sept_85        false  roaring_with_run  avgt    5    2868.225 ±    39.776  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive      wikileaks-noquotes        false  roaring_with_run  avgt    5    9678.845 ±    11.653  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive       census-income_srt        false  roaring_with_run  avgt    5      58.701 ±     1.147  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive          census1881_srt        false  roaring_with_run  avgt    5    4793.926 ±     5.604  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive     weather_sept_85_srt        false  roaring_with_run  avgt    5     788.940 ±     0.329  us/op
RealDataBenchmarkWideOrNaive.wideOr_naive  wikileaks-noquotes_srt        false  roaring_with_run  avgt    5    1887.183 ±     4.304  us/op
```

**Comment by lemire** [comment]
First optimization included in release RoaringBitmap-0.6.17.

Second optimization reverted (reverted also in CRoaring).
